src/lib/FreeC/IR/Inlining.hs

```diff
@@ -76,71 +86,81 @@ inlineExpr decls = inlineAndBind
       ++ " type arguments in an application of '"
       ++ showPretty e
       ++ "'."
-    return (remainingArgs, e')
-
-  -- | Performs inlining on the given subexpression.
-  --
-  --   If a function is inlined, fresh free variables are introduced for the
-  --   function arguments. The first two components of the returned tuple
-  --   contain the names of the type variables and variables that still need
-  --   to be bound. Function application and visible type application
-  --   expressions automatically substitute the corresponding argument for
-  --   the passed value.
-  inlineExpr' :: IR.Expr -> Converter ([String], [String], IR.Expr)
+    return (remainingArgs, e', name)
+
+  -- | Performs inlining on the given subexpression.
































+  --
































+  --   If a function is inlined, fresh free variables are introduced for the
































+  --   function arguments. The first two components of the returned tuple
































+  --   contain the names of the type variables and variables that still need
































+  --   to be bound. The last component of the returndes tuple is the qualified

+  --   name of the inlined function.

+  --   Function application and visible type application






























 expressions

+  --   automatically substitute the corresponding argument for
































+  --   the passed value.
































+  inlineExpr' :: IR.Expr -> Converter ([String], [String], IR.Expr, IR.QName)
   inlineExpr' var@(IR.Var _ name _) = case Map.lookup name declMap of
-    Nothing                    -> return ([], [], var)
+    Nothing                    -> return ([], [], var, emptyName)
     Just (typeArgs, args, rhs) -> do
       (typeArgs', rhs' ) <- renameTypeArgs typeArgs rhs
       (args'    , rhs'') <- renameArgs args rhs'
       return
-        (map IR.typeVarDeclIdent typeArgs', map IR.varPatIdent args', rhs'')
+        ( map IR.typeVarDeclIdent typeArgs'
+        , map IR.varPatIdent      args'
+        , rhs''
+        , name
+        )
```

Comment:
I'm wondering whether it is actually necessary to add an additional return value to `inlineExpr'`. As far as I can tell, you have to return `name` because `inlineAndBind` removes the function with that name from the list of functions to inline in the recursive call. But couldn't the recursive call be performed in here?